File: 프로그래머스/2/17677. ［1차］ 뉴스 클러스터링/［1차］ 뉴스 클러스터링.py
# 자카드 유사도 : 두 집합 A, B => J(A, B) : 교집합 크기 / 합집합 크기
# 두 글자씩 끊어 다중 집합의 원소, 영문으로만 이루어지지 않는 경우 버림
# 대소문자 차이 무시
# 65536
def solution(str1, str2):
    import math
    n = u = 0
    check1 = dict()
    check2 = dict()
    for i in range(len(str1) - 1):
        if str1[i:i+2].isalpha():
            check1.setdefault(str1[i:i+2].lower(), 0)
            check1[str1[i:i+2].lower()] += 1
    for i in range(len(str2) - 1):
        if str2[i:i+2].isalpha():
            # 이미 값이 있는 친구들을 더해줄 필요가 있는지.. 있지.. 
            check2.setdefault(str2[i:i+2].lower(), 0)
            check2[str2[i:i+2].lower()] += 1
    print(check1, check2)
    for i in check1.keys():
        # 둘 다에 존재한다면 => 더 큰 값으로 합집합에 더해주기 , 교집합
        if i in check2.keys():
            u += max(check1[i], check2[i])
            n += min(check1[i], check2[i])
        # 하나에만 존재한다면 => 합집합에 더해주기
        else:
            u += check1[i]
    for i in check2.keys():
        if i not in check1.keys():
            u += check2[i]
    if u == 0:
        return 65536
    else:
        return math.floor(n/u*65536)

File: 프로그래머스/2/17677. ［1차］ 뉴스 클러스터링/［1차］ 뉴스 클러스터링.java
// 교집합 크기 / 합집합 크기
// 대소문자 차이 무시. 두 글자씩 끊어 다중 집합의 원소, 영문으로만 이루어지지 않는 경우 버림
// 자카드 유사도에 * 65536 소수점 아래 버림
import java.util.*;
class Solution {
    public int solution(String str1, String str2) {
        float n = 0, u = 0; // 교집합, 합집합 크기
        Map<String, Integer> map1 = new HashMap<>(); // 첫 번째 문자열 조합
        Map<String, Integer> map2 = new HashMap<>(); // 두 번째 문자열 조합
        // str1 순회
        for (int i = 0; i < str1.length() - 1; i++){
            if (str1.substring(i, i+2).matches("^[a-zA-Z]*$")) {
                String word = str1.substring(i, i+2).toLowerCase();
                map1.putIfAbsent(word, 0);
                map1.replace(word, map1.get(word) + 1);
            }
        }
        // str2 순회
        for (int j = 0; j < str2.length() - 1; j++){
            if (str2.substring(j, j+2).matches("^[a-zA-Z]*$")) {
                String word = str2.substring(j, j+2).toLowerCase();
                map2.putIfAbsent(word, 0);
                map2.replace(word, map2.get(word) + 1);
            }
        }
        // key 순회
        for (String key : map1.keySet()){
            // 둘 모두에 포함된다면 교집합++, 아니라면 합집합++
            if(map2.keySet().contains(key)) {
                u += Math.max(map1.get(key), map2.get(key));
                n += Math.min(map1.get(key), map2.get(key));
            } else {
                u += map1.get(key);
            }
        }
        for (String key : map2.keySet()){
            if(!map1.keySet().contains(key)) {
                u += map2.get(key);
            }
        }
        
        if (u == 0) {
            return 65536;
        } else {
            return (int) Math.floor(n / u * 65536);
        }
    }
}